`assert a == approx(b)` when `b` is dict containing zero value results in ZeroDivisionError

Pytest behaves differently when comparing dictionaries containing zero values compared to lists containing same values

pytest==7.2.0
Ubuntu 22.04

```python
import pytest

def test_foo_dict():
    a = {'foo': 42.0}
    b = {'foo': 0.0}
    assert a == pytest.approx(b) # ZeroDivisionError in pytest/python_api.py

def test_foo_list():
    a = [42.0]
    b = [0.0]
    assert a == pytest.approx(b) # OK
```

```python
_____________________ test_foo_dict

    def test_foo_dict():
        a = {'foo': 42.0}
        b = {'foo': 0.0}
>       assert a == pytest.approx(b)
E       AssertionError: assert {'foo': 42.0} == approx({'foo': 0.0 ± 1.0e-12})
E         (pytest_assertion plugin: representation of details failed: /home/arkanoid/test/venv/lib/python3.10/site-packages/_pytest/python_api.py:274: ZeroDivisionError: float division by zero.
E          Probably an object has a faulty __repr__.)

extra/test_pytest_issue.py:9: AssertionError

_____________________ test_foo_list

    def test_foo_list():
        a = [42.0]
        b = [0.0]
>       assert a == pytest.approx(b)
E       assert [42.0] == approx([0.0 ± 1.0e-12])
E         comparison failed. Mismatched elements: 1 / 1:
E         Max absolute difference: 42.0
E         Max relative difference: 1.0
E         Index | Obtained | Expected     
E         0     | 42.0     | 0.0 ± 1.0e-12

extra/test_pytest_issue.py:15: AssertionError
```

--- a/changelog/10533.bugfix.rst
+++ b/changelog/10533.bugfix.rst
@@ -1,0 +1,1 @@
+Fix :func:`pytest.approx` handling of dictionaries containing one or more values of `0.0` in class ApproxMapping.

--- a/src/_pytest/python_api.py
+++ b/src/_pytest/python_api.py
@@ -266,14 +266,17 @@ class ApproxMapping(ApproxBase):
             approx_side_as_map.items(), other_side.values()
         ):
             if approx_value != other_value:
-                max_abs_diff = max(
-                    max_abs_diff, abs(approx_value.expected - other_value)
-                )
-                max_rel_diff = max(
-                    max_rel_diff,
-                    abs((approx_value.expected - other_value) / approx_value.expected),
-                )
-                different_ids.append(approx_key)
+              max_abs_diff = max(
+                  max_abs_diff, abs(approx_value.expected - other_value)
+              )
+              if approx_value.expected == 0.0:
+                  max_rel_diff = math.inf
+              else:
+                  max_rel_diff = max(
+                      max_rel_diff,
+                      abs((approx_value.expected - other_value) / approx_value.expected),
+                  )
+              different_ids.append(approx_key)
 
         message_data = [
             (str(key), str(other_side[key]), str(approx_side_as_map[key]))

--- a/src/_pytest/python_api.py
+++ b/src/_pytest/python_api.py
@@ -266,17 +266,20 @@ class ApproxMapping(ApproxBase):
             approx_side_as_map.items(), other_side.values()
         ):
             if approx_value != other_value:
-              max_abs_diff = max(
-                  max_abs_diff, abs(approx_value.expected - other_value)
-              )
-              if approx_value.expected == 0.0:
-                  max_rel_diff = math.inf
-              else:
-                  max_rel_diff = max(
-                      max_rel_diff,
-                      abs((approx_value.expected - other_value) / approx_value.expected),
-                  )
-              different_ids.append(approx_key)
+                max_abs_diff = max(
+                    max_abs_diff, abs(approx_value.expected - other_value)
+                )
+                if approx_value.expected == 0.0:
+                    max_rel_diff = math.inf
+                else:
+                    max_rel_diff = max(
+                        max_rel_diff,
+                        abs(
+                            (approx_value.expected - other_value)
+                            / approx_value.expected
+                        ),
+                    )
+                different_ids.append(approx_key)
 
         message_data = [
             (str(key), str(other_side[key]), str(approx_side_as_map[key]))

--- a/testing/python/approx.py
+++ b/testing/python/approx.py
@@ -630,6 +630,19 @@ class TestApprox:
     def test_dict_vs_other(self):
         assert 1 != approx({"a": 0})
 
+    def test_dict_for_div_by_zero(self, assert_approx_raises_regex):
+        assert_approx_raises_regex(
+            {"foo": 42.0},
+            {"foo": 0.0},
+            [
+                r"  comparison failed. Mismatched elements: 1 / 1:",
+                rf"  Max absolute difference: {SOME_FLOAT}",
+                r"  Max relative difference: inf",
+                r"  Index \| Obtained\s+\| Expected   ",
+                rf"  foo   | {SOME_FLOAT} \| {SOME_FLOAT} ± {SOME_FLOAT}",
+            ],
+        )
+
     def test_numpy_array(self):
         np = pytest.importorskip("numpy")
 

